### DISCUSSION TITLE
Erasing NVS memory

### DISCUSSION THREAD

**Discussion by grzegorzwozny** [micropython/micropython#11023] [ESP32]
Hello,
How to erasing [Non-Volatile Storage](https://docs.micropython.org/en/v1.15/library/esp32.html#non-volatile-storage) memory in ESP32 using MicroPython?

I [store](https://forum.micropython.org/viewtopic.php?t=10230#p56911) data in NVS as `key:value` in namespace:

```
namespace1 {
  key1:value1
  key2:value2
}
namespace2 {
  key:value2
}
```

In C/C++ is it easy to do using [nvs_flash.h](https://github.com/espressif/esp-idf/blob/master/components/nvs_flash/include/nvs_flash.h#L134) module:

```
#include <nvs_flash.h>

void setup() {
  nvs_flash_erase(); // erase the NVS partition and...
  nvs_flash_init(); // initialize the NVS partition.
  while(true);
}

void loop() {

}
```

In class `esp32.NVS` I see only one function `NVS.erase_key(key)` to erase single key but how to erase all namespaces and keys?

**Comment by grzegorzwozny** [comment]
At the moment, I see only one option to erase the NVS partition. Based on the address and size information of the [NVS partition](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/partition-tables.html#built-in-partition-tables):

```
# ESP-IDF Partition Table
# Name,   Type, SubType, Offset,  Size, Flags
nvs,      data, nvs,     0x9000,  0x4000,
otadata,  data, ota,     0xd000,  0x2000,
phy_init, data, phy,     0xf000,  0x1000,
factory,  app,  factory, 0x10000,  1M,
ota_0,    app,  ota_0,   0x110000, 1M,
ota_1,    app,  ota_1,   0x210000, 1M,
```

I can use [esptool](https://docs.espressif.com/projects/esptool/en/latest/esp32s3/esptool/basic-commands.html#erase-flash-erase-flash-erase-region) to achieve this, like this:
`esptool.py erase_region 0x9000 0x4000`

But this is not a solution. I looking for a function for do this...

**Reply by karfas** [reply]
> But this is not a solution. I looking for a function for do this...

You can do the same thing, I think, from within micropython using `Partition.writeblocks()`. 

See https://docs.micropython.org/en/latest/library/esp32.html; some example usage comes with micropython in [esp32/modules](https://github.com/micropython/micropython/tree/master/ports/esp32/modules).

**Reply by FeroxTL** [reply]
```python
import esp32
import machine

p = esp32.Partition.find(esp32.Partition.TYPE_DATA, label='nvs')[0]

# p.info()[3] is partition size
for x in range(int(p.info()[3] / 4096)):
    p.writeblocks(x, bytearray(4096))

machine.reset()
```

This would completely erase NVS partition. `machine.reset` is required to reinitialize micropython structures
